### PR TITLE
Ncgenerics test fixes kavon v10

### PR DESCRIPTION
**Pull request by kavon**
more test fixes

### CODE REVIEW

**Comment by kavon** [comment]
preset=ncgenerics,smoketest=macosx
@swift-ci Please test with preset macOS Platform

**Comment by kavon** [comment]
@swift-ci smoke test

**Comment by kavon** [comment]
preset=ncgenerics,smoketest=macosx
@swift-ci Please test with preset macOS Platform

File: lib/AST/ConformanceLookup.cpp [hunk 1]
```diff
@@ -586,6 +586,20 @@ LookupConformanceInModuleRequest::evaluate(
...
     return ProtocolConformanceRef::forMissingOrInvalid(type, protocol);
 
+  // We specially avoid recording conformances to invertible protocols in a
+  // class's conformance table. This prevents an evaulator cycle.
+  if (ctx.LangOpts.hasFeature(Feature::NoncopyableGenerics) &&
+      isa<ClassDecl>(nominal)) {
+    if (auto ip = protocol->getInvertibleProtocolKind()) {
+      ImplicitKnownProtocolConformanceRequest req{nominal,
+                                                  getKnownProtocolKind(*ip)};
+      if (auto conformance = evaluateOrDefault(ctx.evaluator, req, nullptr))
```
Comment (slavapestov) [inline lib/AST/ConformanceLookup.cpp:596]: Can the request just return a ProtocolConformanceRef

File: lib/AST/GenericEnvironment.cpp [hunk 2]
```diff
@@ -417,6 +417,9 @@ Type GenericEnvironment::mapTypeIntoContext(GenericEnvironment *env,
 }
 
 Type MapTypeOutOfContext::operator()(SubstitutableType *type) const {
+  if (auto genericParam = dyn_cast<GenericTypeParamType>(type))
```
Comment (slavapestov) [inline lib/AST/GenericEnvironment.cpp:420]: Bug

File: lib/Sema/TypeCheckInvertible.cpp [hunk 3]
```diff
@@ -349,16 +349,26 @@ ProtocolConformance *deriveConformanceForInvertible(Evaluator &evaluator,
...
+          ConformanceEntryKind::Synthesized, nullptr);
+
+      nominal->registerProtocolConformance(conformance, /*synthesized=*/true);
+      return conformance;
+    } else {
+      // Classes need to return a built-in conformance to avoid cycles with
+      // superclasses, such as:
+      //    class Base<T>: IP {}
+      //    class Derived: Base<Derived> {}
+      return ctx.getBuiltinConformance(nominal->getDeclaredInterfaceType(),
```
Comment (slavapestov) [inline lib/Sema/TypeCheckInvertible.cpp:368]: the conformance should be formed in lookupConformance() with the conforming type, not the nominal's declared interface type. Generally we want this to hold:

```
lookupConformance(type, proto)->getType()->isEqual(type)
```

File: lib/Sema/TypeCheckProtocol.cpp [hunk 4]
```diff
@@ -4864,6 +4864,11 @@ static void ensureRequirementsAreSatisfied(ASTContext &ctx,
       auto assocConf = conformance->getAssociatedConformance(depTy, proto);
       if (assocConf.isConcrete()) {
         auto *concrete = assocConf.getConcrete();
+
+        // Builtin conformances are always available
+        if (isa<BuiltinProtocolConformance>(concrete))
+          return false;
+
         auto replacementTy = dc->mapTypeIntoContext(concrete->getType());
```
Comment (slavapestov) [inline lib/Sema/TypeCheckProtocol.cpp:4872]: You won't need the `if` above once you fix the `getType()` invariant

**Comment by kavon** [comment]
@swift-ci smoke test

**Comment by kavon** [comment]
@swift-ci smoke test

**Pull request merged**